include/revng/Model/Binary.h

```diff
@@ -162,66 +191,123 @@ class model::Binary : public model::generated::Binary {
       static_assert(std::is_rvalue_reference_v<decltype(NewType)>);
-      revng_assert(NewType->ID() != 0);
       Inserter.emplace(std::move(NewType));
     }
   }
 
-  template<derived_from<model::TypeDefinition> NewType,
-           typename... ArgumentTypes>
-  [[nodiscard]] std::pair<NewType &, model::DefinitionReference>
-  makeTypeDefinition(ArgumentTypes &&...Arguments) {
-    using UT = model::UpcastableTypeDefinition;
-    UT Result = UT::make<NewType>(std::forward<ArgumentTypes>(Arguments)...);
-    model::DefinitionReference ResultPath = recordNewType(std::move(Result));
-    return { *llvm::cast<NewType>(ResultPath.get()), ResultPath };
+public:
+  /// \note Only use this when absolutely necessary, for example, when doing
+  ///       bulk reference replacement.
+  ///       In the general case prefer \ref makeType instead.
+  model::DefinitionReference
+  getDefinitionReference(const model::TypeDefinition::Key &Key) {
+    return DefinitionReference::fromString(this,
+                                           "/TypeDefinitions/"
+                                             + getNameFromYAMLScalar(Key));
+  }
+  model::DefinitionReference
+  getDefinitionReference(const model::TypeDefinition::Key &Key) const {
+    return DefinitionReference::fromString(this,
+                                           "/TypeDefinitions/"
+                                             + getNameFromYAMLScalar(Key));
   }
 
-  model::DefinitionReference getPrimitiveType(PrimitiveKind::Values V,
-                                              uint8_t ByteSize);
-
-  model::DefinitionReference getPrimitiveType(PrimitiveKind::Values V,
-                                              uint8_t ByteSize) const;
+  model::UpcastableType makeType(const model::TypeDefinition::Key &Key) {
+    return model::DefinedType::make(getDefinitionReference(Key));
+  }
+  model::UpcastableType makeConstType(const model::TypeDefinition::Key &Key) {
+    return model::DefinedType::makeConst(getDefinitionReference(Key));
+  }
+  model::UpcastableType makeType(const model::TypeDefinition::Key &Key) const {
+    return model::DefinedType::make(getDefinitionReference(Key));
+  }
+  model::UpcastableType
+  makeConstType(const model::TypeDefinition::Key &Key) const {
+    return model::DefinedType::makeConst(getDefinitionReference(Key));
+  }
 
-  bool verifyTypeDefinitions() const debug_function;
-  bool verifyTypeDefinitions(bool Assert) const debug_function;
-  bool verifyTypeDefinitions(VerifyHelper &VH) const;
+  /// Return the first available type ID
+  uint64_t getAvailableTypeID() const;
 
 public:
-  std::string path(const model::Function &F) const {
-    return "/Functions/" + key(F);
+  /// The helper for the prototype unwrapping.
+  /// Use this when you need to access/modify the existing prototype,
+  /// and \ref DefaultPrototype() when you need to assign a new one.
+  model::TypeDefinition *defaultPrototype() {
```

Comment:
Not sure how useful this is, you still need an `if` in the caller.

Reply:
There are `raw` and `cabi` versions if you want to skip the one of the ifs. This is provided to be used as a fallback by the prototype selecting helpers (ones that check for call sight prototypes).